presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java

```diff
@@ -174,19 +201,22 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
         return Futures.catchingAsync(resultFuture, Exception.class, e -> {
             if (e instanceof InterruptedException) {
                 Thread.currentThread().interrupt();
-                throw propagate(e);
+                throw propagate(retryStatus, stats, e);
             }
             else if (retryStatus.shouldStopRetry(e)) {
-                throw propagate(e);
+                throw propagate(retryStatus, stats, e);
             }
             log.warn(e, "Failed on executing %s with attempt %d, will perform async retry.", callableName, retryStatus.getAttempts());
-            return Futures.dereference(retryExecutorService.schedule(() -> runAsyncInternal(callableName, callable, retryStatus), retryStatus.getRetryDelayInMs(), MILLISECONDS));
+            return Futures.dereference(retryExecutorService.schedule(() -> runAsyncInternal(callableName, stats, callable, retryStatus), retryStatus.getRetryDelayInMs(), MILLISECONDS));
         });
     }
 
-    private RuntimeException propagate(Exception e)
+    private RuntimeException propagate(RetryStatus retryStatus, RetryStats stats, Exception e)
     {
         Exception classified = classifier.apply(e);
+        stats.addFinalFailure();
```

Comment:
This is really unrelated to throwing an exception. How about we separate it into a separate method `recordFinalFailure(RetryStatus, RetryStats)`

Reply:
Also, it's weird to shove this in the middle of where `classified` is computed and used